sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala

```diff
@@ -20,19 +20,347 @@ package org.apache.spark.sql.catalyst.optimizer
+              val joinCond = findJoinConditions(factTable, plan, conditions)
+              joinCond.nonEmpty
+            }
+
+            // Find the corresponding join conditions.
+            val allFactJoinCond = allFactJoins.flatMap { plan =>
+              val joinCond = findJoinConditions(factTable, plan, conditions)
+              joinCond
+            }
+
+            // Verify if the join columns have valid statistics
+            val areStatsAvailable = allFactJoins.forall { dimTable =>
+              allFactJoinCond.exists {
+                case BinaryComparison(lhs: AttributeReference, rhs: AttributeReference) =>
+                  val dimCol = if (dimTable.outputSet.contains(lhs)) lhs else rhs
+                  val factCol = if (factTable.outputSet.contains(lhs)) lhs else rhs
+                  hasStatistics(dimCol, dimTable) && hasStatistics(factCol, factTable)
+                case _ => false
+              }
+            }
+
+            if (!areStatsAvailable) {
+              emptyStarJoinPlan
+            } else {
+              // Find the subset of dimension tables. A dimension table is assumed to be in
+              // RI relationship with the fact table. Also, only consider equi-joins
+              // between a fact and a dimension table.
+              val eligibleDimPlans = allFactJoins.filter { dimTable =>
+                allFactJoinCond.exists {
+                  case cond @ BinaryComparison(lhs: AttributeReference, rhs: AttributeReference)
+                    if cond.isInstanceOf[EqualTo] || cond.isInstanceOf[EqualNullSafe] =>
+                    val dimCol = if (dimTable.outputSet.contains(lhs)) lhs else rhs
+                    isUnique(dimCol, dimTable)
+                  case _ => false
+                }
+              }
+
+              if (eligibleDimPlans.isEmpty) {
+                // An eligible star join was not found because the join is not
+                // an RI join, or the star join is an expanding join.
+                emptyStarJoinPlan
+              } else {
+                Seq(factTable +: eligibleDimPlans)
+              }
+            }
+        }
+      }
+    }
+  }
+
+  /**
+   * Reorders a star join based on heuristics:
+   *   1) Finds the star join with the largest fact table and places it on the driving
+   *      arm of the left-deep tree. This plan avoids large table access on the inner, and
+   *      thus favor hash joins.
+   *   2) Applies the most selective dimensions early in the plan to reduce the amount of
+   *      data flow.
+   */
+  def reorderStarJoins(
+      input: Seq[(LogicalPlan, InnerLike)],
+      conditions: Seq[Expression]): Seq[(LogicalPlan, InnerLike)] = {
+    assert(input.size >= 2)
+
+    val emptyStarJoinPlan = Seq.empty[(LogicalPlan, InnerLike)]
+
+    // Find the eligible star plans. Currently, it only returns
+    // the star join with the largest fact table.
+    val eligibleJoins = input.collect{ case (plan, Inner) => plan }
```

Comment:
@cloud-fan To provide more context about my above reply.  My initial algorithm kept track of both (plan, InnerLike) since that's what ```JoinReorder``` expects. To interface with CBO, I generalized the algorithm to only handle logical plans, since that's what ```CostBasedReorder``` expects. The latter is a more intuitive plan manipulation. I don't think it's worth keeping two interfaces, and therefore, I decided to simply manipulate the Inner plans, which is what I would expect to see in a star-join query. But if you think it's an issue that we need to address, I can think of some mapping between the two. But I would prefer to keep it simple.